src/common/templates/_deployment-lifecycle.tpl

```diff
@@ -0,0 +1,80 @@
+{{/*
+USAGE:
+{{- include "harnesscommon.v1.renderLifecycleHooks" (dict "ctx" $) }}
+A JFR hook can be automatically added if JFR is enabled
+*/}}
+{{- define "harnesscommon.v1.renderLifecycleHooks" }}
+{{- $ := .ctx }}
+{{- if $.Values.lifecycleHooks }}
+{{ include "harnesscommon.tplvalues.render" (dict "value" $.Values.lifecycleHooks "context" $) }}
+{{- else if $.Values.global.jfr.enabled }}
+postStart:
+  exec:
+    command:
+    - /bin/sh
+    - -c
+    - |
+      mkdir -p ${JFR_DUMP_ROOT_LOCATION}/dumps/${SERVICE_NAME}/${ENV_TYPE}/jfr_dumps/${POD_NAME};
+      ln -s ${JFR_DUMP_ROOT_LOCATION}/dumps/${SERVICE_NAME}/${ENV_TYPE}/jfr_dumps/${POD_NAME} ${JFR_DUMP_ROOT_LOCATION}/POD_NAME ;
+preStop:
+  exec:
+    command:
+    - /bin/sh
+    - -c
+    - |
+      ## The pod is out of service at this point in Endpoints (nominally).  Allow time before we start signally process
+      ## stops and termination via the shutdown file.
+      sleep {{- default "30" (default dict $.Values.lifecycleHooks).terminationTime }};
+      ts=$(date '+%s');
+      loc=${JFR_DUMP_ROOT_LOCATION}/dumps/${SERVICE_NAME}/${ENV_TYPE}/$ts/${POD_NAME};
+      mkdir -p $loc; sleep 1; echo $ts > $loc/restart;
+      echo $(date '+%s') > $loc/begin;
+      PID=$(jps|grep -vi jps|awk '{ print $1}');
+
+      #Copy GC log file
+      cp mygclogfilename.gc $loc/;
+
+      #Retry 10 times to take thread dump. Unsuccessful attempt has just 1 line with java process id in the output.
+      for ((n=0;n<10;n++)); do
+      jcmd $PID Thread.print -e > $loc/thread-dump-attempt-$n.txt;
+      if [ $(wc -l < $loc/thread-dump-attempt-$n.txt) -gt 1 ]; then break; fi;
+      done
+
+      #10 retries to take heap histogram. Unsuccessful attempt has just 1 line with java process id in the output.
+      for ((n=0;n<10;n++)); do
+      jcmd $PID GC.class_histogram -all > $loc/heap-histogram-attempt-$n.txt;
+      if [ $(wc -l < $loc/heap-histogram-attempt-$n.txt) -gt 1 ]; then break; fi;
+      done
+
+      ## Dump native memory for analysis
+      jcmd $PID VM.native_memory  > $loc/native-memory-dump.txt;
+
+      #Dump latest chunk of JFR recording
+      jcmd $PID JFR.dump name=jfrRecording filename=${JFR_DUMP_ROOT_LOCATION}/dumps/${SERVICE_NAME}/${ENV_TYPE}/jfr_dumps/${POD_NAME}/container_termination_$(date +%Y_%m_%d_%H_%M_%S).jfr > $loc/jfr_done.txt
+
+      echo $(date '+%s') > $loc/end
+
+      ## Once JFR is caught, drop maintenace file & sleep final for grace period & complete termination
+      touch shutdown;
+      sleep {{- default "30" (default dict $.Values.lifecycleHooks).terminationGraceTime }};
+      ## Send SIGTERM which starts JVM Shutdown hooks & upon completion will exit
+      kill -15 $PID;
+{{- else }}
```

Comment:
this might add unnecessary wait time of 60s in every service.

Reply:
SO anything with a GRPC server needs this.  Further, this allows full time for any service to cleanly STOP background processing.  Yes it's intentional to all services.  NOTE since we turn on JFR on all services, that is ALREADY happening - this applies to defaults for places that's missed (e.g. on-premise).  I'd like to reduce it in the future.  This CAN be disabled by passing custom lifecycle hooks, aka this is a "saner" default.

Reply:
on-premise is whats the concern here. SMP customers already have complain that upgrades take 15+ minutes due to inter-service dependency and slowness of services to come online. With a default wait time of 60 seconds for every pod, upgrade will become slower. 
I would recommend we add support for a default global lifecycle hook which we can update in our internal env overrides for now.

Reply:
Just pushed one change to see if lifecycle exists and is null to blank out the entire value instead.